### PR TITLE
Centralize test setup

### PR DESCRIPTION
**Pull request by snejus**
This PR deduplicates shared tests setup that was used across test files. It attempts to centralise the setup into a single source of truth implementation in `beets.test.helper`.

This works towards the eventual migration to `pytest` (#5361) making it easier to replace the tests setup with `pytest` fixtures.

It builds upon the temporary files cleanup in #5345.

### Test setup

* Mostly duplicate test setup implementations in `beets.test._common.TestCase` and `beets.test.helper.TestHelper` were merged into a single implementation as `beets.test.helper.BeetsTestCase`.
* Replaced `unittest.TestCase` and `beets.test.helper.TestHelper` combination with `beets.test.helper.ImportTestCase`
* Refactored `test/test_plugins.py` removing duplicate import files setup.
* Centralised setting up the database on disk.
* Centralised plugin test setup into `beets.test.helpers.PluginMixin`.

### Removals

* Around 1/3 of the removed lines correspond to the removal of `def suite()` functions that were previously used for tests collection.
* Removed redundant `setUp` and `tearDown` methods from many test files given that the functions that they performed had been centralised in `beets.test.helper`.
* Removed redundant library initialisations

### Importing

* Centralised import directory creation (only gets created on the first access).
* Deduplicated `_setup_import_session` implementations in `TerminalImportSessionSetup` and `ImportHelper`.
* Merged `ImportHelper._create_import_dir` and `TestHelper.create_importer` implementations into `ImportHelper.setup_importer`.
* Merged various takes on import files prep into `ImportHelper.prepare_albums_for_import` and `ImportHelper.prepare_album_for_import`.
* Seeing that many tests used it, defined `AsIsImporterMixin` which provides a method `run_asis_importer` to setup _asis_ (non-autotag) importer and run it

### CODE REVIEW

File: test/test_plugins.py
Comment (bal-e) [inline test/test_plugins.py:227]: This definitely looks better than using an f-string, but I can't wait for that to change with `pathlib` integration.

File: test/plugins/test_filefilter.py
Comment (bal-e) [inline test/plugins/test_filefilter.py:103]: `__run` has been renamed to `_run`, but then shouldn't `__create_import_dir` also be renamed to `_create_import_dir`?

File: test/plugins/test_replaygain.py
Comment (bal-e) [inline test/plugins/test_replaygain.py:69]: Why not use `PluginTestCase` here?  And why do you have to manually call `tearDown()` if an exception occurs? (won't `pytest` do that for you automatically?)

**Review by bal-e — COMMENTED** [review]
This is an incredibly comprehensive PR!  I've left a few notes, but everything looks good.  Perhaps you could use `pathlib` in the functions you've rewritten, so that they don't need to be migrated later?  Or we can leave that for the later `pathlib`-specific PR(s).

**Comment by Serene-Arc** [comment]
I'll try and get my review in over the next day or two! I want to really sit down and go through it all.

File: test/plugins/test_filefilter.py
Comment (snejus) [reply test/plugins/test_filefilter.py]: Realised I missed out on removing the import files setup from here. Ended up refactoring this module significantly - the aforementioned method is now gone and the tests should be more clear to follow.

File: test/plugins/test_embedart.py
Comment (snejus) [reply test/plugins/test_embedart.py:49]: This one's an interesting one - it doesn't seem to run `setup_beets` and thus doesn't seem to require the entire temporary directory / lib setup.

I think it only needs a few methods that `TestHelper` provides. Though now that you brought this to attention, I can see that this test class is actually being ignored:

```python
class EmbedartCliTest(TestHelper, FetchImageHelper):
```

It doesn't inherit from `unittest.TestCase`, so it doesn't get picked up. Will sort this out and check whether we have more cases like this one.

File: test/test_plugins.py
Comment (snejus) [reply test/test_plugins.py:227]: > I can't wait for that to change with `pathlib` integration.

Oh yes, completely. This will remove amounts of boilerplate from `beets.util`.

File: test/plugins/test_replaygain.py
Comment (snejus) [reply test/plugins/test_replaygain.py:69]: You will see there are a more test modules which run `load_plugins` and `unload_plugins` manually and they all have something in common: they set up plugin configuration before loading it.

Some of them test various configurations, like `test/plugin/test_filtefilter.py`, where the plugin is loaded midway through the test.

`PluginMixin / PluginTestCase` loads the plugin at the end of the setup, before the test methods runs, therefore plugin configuration defined in the test method simply has no effect. This mixin is only useful for those plugin tests that simply need the plugin to be loaded (majority of them). Otherwise, as you see here, the plugin is loaded manually.

I think it may be a good idea to define a method `PluginMixin.reload_plugin` which does something like

```python
class PluginMixin:
    ...
    def reload_plugin(self, config: dict):
        self.unload_plugins()
        beets.config[self.plugin].update(config)
        self.load_plugins(self.plugin)
```

**Comment by snejus** [comment]
> Perhaps you could use `pathlib` in the functions you've rewritten, so that they don't need to be migrated later? Or we can leave that for the later `pathlib`-specific PR(s).

That's a very good point. I was careful to not try and refactor anything to do with paths (out of scope here, and the PR size would have quickly gotten out of hand) but I think it's a good idea to use `pathlib` in rewrites/new stuff.

For now I made the new methods in `ImportHelper` to use `pathlib` and you can see it's already made it cleaner!

File: test/plugins/test_embedart.py
Comment (bal-e) [reply test/plugins/test_embedart.py:49]: It might not need the additional setup, but it might be worth it to inherit from `BeetsTestCase` just for consistency with other tests.  Otherwise, there should be a comment explaining why `BeetsTestCase` should _not_ be inherited from.

File: test/plugins/test_replaygain.py
Comment (bal-e) [reply test/plugins/test_replaygain.py:69]: Right, thank you for explaining!  But would a `test_plugin_load()` method (set config; load plugin; unload plugin) be more appropriate, if tests call `reload_plugin()` to try loading a plugin temporarily?

**Comment by bal-e** [comment]
> That's a very good point. I was careful to not try and refactor anything to do with paths (out of scope here, and the PR size would have quickly gotten out of hand) but I think it's a good idea to use `pathlib` in rewrites/new stuff.

I agree.  When we fully migrate to `pytest`, we can use its tempdir fixtures which already use `pathlib`.  That should also end up migrating the majority of the tests to `pathlib`.

File: test/plugins/test_embedart.py
Comment (snejus) [reply test/plugins/test_embedart.py:49]: See #5363 where I enabled this test and a couple of others that are skipped.

Comment (snejus) [reply test/plugins/test_embedart.py:49]: @bal-e, maybe it would be a good idea to move the the functionality that this test needs away from `BeetsTestCase` to a `LibMixin` (or similar)? There are a couple of other tests (namely `test_query.py`) that are based from `BeetsTestCase` but they actually do not require the entire temporary dir setup which slows them down slightly.

File: test/plugins/test_replaygain.py
Comment (snejus) [reply test/plugins/test_replaygain.py:69]: @bal-e I'm not sure whether I understood what you meant correctly. Maybe something like

```python
from contextlib import contextmanager

...

class PluginMixin:
    @contextmanager
    def configure_plugin(self, config):
        beets.config[self.plugin].update(config)
        self.load_plugins(self.plugin)

        yield

        self.unload_plugins()

class PluginTest(PluginMixin, ...):
    def test_plugin_config(self):
        with self.configure_plugin({"option": "value"}):
            self.assertEqual(beets.config[self.plugin]["option"].get(), "value")
```

This would clean up the configuration on test exit.

File: test/plugins/test_embedart.py
Comment (snejus) [reply test/plugins/test_embedart.py:50]: Ignore my message above - it actually requires the entire setup! I have now made it inherit from `BeetsTestCase`.

File: test/test_importer.py
Comment (bal-e) [inline test/test_importer.py:1705]: Why have you made the log list a set?  Doesn't that obscure the order of the logs and whether they were produced multiple times?

**Comment by bal-e** [comment]
I like the new uses of `pathlib` and f-strings, the code feels a lot cleaner.  It also looks like the net line change is becoming even more negative, which is awesome.  Good job @snejus!

File: test/plugins/test_mbsubmit.py
Comment (Serene-Arc) [inline test/plugins/test_mbsubmit.py:52]: Double checking here, what's the reason for this change? We're changing what the test compares to, which means it should have been failing before if there's no change to the program code. If it wasn't, then that's an internal rule in our tests that should be marked.

File: test/plugins/test_replaygain.py
Comment (Serene-Arc) [reply test/plugins/test_replaygain.py:69]: I'll point out that this context manager approach is very, very similar to the pytest approach with fixtures. If we do that, it'll be less of a change when converting the remainder of the files.

File: beets/test/helper.py
Comment (Serene-Arc) [inline beets/test/helper.py:546]: I had a big comment written out but then I looked it up and PEP 604 was backported through the futures. I'll have to update my typing habits. Great that you're using the more modern way, we'll have to make the change through the codebase (or at least the parts with typing).

File: test/test_importer.py
Comment (Serene-Arc) [reply test/test_importer.py:1705]: I don't think order should matter in this specific part, but the duplicate entry issue is something we should test for.

**Review by Serene-Arc — CHANGES_REQUESTED** [review]

**Comment by Serene-Arc** [comment]
I think it would be good if I opened a pathlib PR (as I keep saying I will) after this PR is done. Like this PR, it will have a lot of small line changes that are just replacing function calls and simplifying, but this PR being merged will make that shorter and easier to parse.

File: test/plugins/test_replaygain.py
Comment (snejus) [reply test/plugins/test_replaygain.py]: @Serene-Arc completely agree. I have done this in 28ae414.

File: test/test_importer.py
Comment (snejus) [reply test/test_importer.py:1703]: Since I made the import dir to use the import directory as it is, it didn't seem possible to predict whether the album or the single gets picked up first (sometimes it's former and sometimes it's the latter).

Comment (snejus) [reply test/test_importer.py:1703]: @Serene-Arc I think we could assert the number of imported items/albums to verify that duplicate imports are not happening. See `test_filefilter` where I replaced the logs assertions by library assertions.

File: test/plugins/test_mbsubmit.py
Comment (snejus) [reply test/plugins/test_mbsubmit.py:52]: See 5643358. We had two similar ways how the import files were being prepared, one of them used **Title** in the item title, the other one used **Track**. I found that **Track** is used in more tests, so I removed the other one and updated the tests accordingly.

**Comment by snejus** [comment]
> I think it would be good if I opened a pathlib PR (as I keep saying I will) after this PR is done. Like this PR, it will have a lot of small line changes that are just replacing function calls and simplifying, but this PR being merged will make that shorter and easier to parse.

Sounds like a good plan!

**Comment by snejus** [comment]
I changed this PR to be based on #5363, which is in turn based on #5345:

#5345 <- #5363 <- #5362.

#5345 is ready and can be merged once approved.
#5363 needs a bit more work regarding the windows runner. I will be away from today until 28th so I will come back to it then. Once #5363 is merged, this one follows!

File: test/test_importer.py
Comment (Serene-Arc) [reply test/test_importer.py:1703]: That would be a good solution I think, it would also skip the unlikely situation where the logs are wrong, which would be a separate issue but it's better to make sure that there's actually no duplicate imports.

File: test/plugins/test_mbsubmit.py
Comment (Serene-Arc) [reply test/plugins/test_mbsubmit.py:52]: Cool, thanks for explaining that.

**Comment by Serene-Arc** [comment]
Ah, great. I'll try and get a review in for #5345 over the next day or two so we can get that merged and get the ball rolling on these PRs.

**Comment by snejus** [comment]
> I changed this PR to be based on #5363, which is in turn based on #5345:
> 
> #5345 <- #5363 <- #5362.
> 
> #5345 is ready and can be merged once approved. #5363 needs a bit more work regarding the windows runner. I will be away from today until 28th so I will come back to it then. Once #5363 is merged, this one follows!

@Serene-Arc actually ignore this - I'm not sure what made me to base this PR off #5363 as it actually does not depend on any changes from there. I now rebased this PR off master, so it's ready to be reviewed and tests should be happy.

File: test/plugins/test_hook.py
Comment (bal-e) [inline test/plugins/test_hook.py:61]: Can we implement this as `self.assertIn` instead?  What does the error message contained here look like, and how does it vary?  If this style is necessary, a comment explaining the variation in the message is needed.

**Review by bal-e — APPROVED** [review]
This is looking really good!  I've outlined a few nitpicks, but overall this is awesome.  @snejus, I love the amount of code clean-up you've managed in this ginormous PR, good job!

File: test/test_importer.py
Comment (snejus) [reply test/test_importer.py:1705]: > Since I made the import dir to use the import directory as it is, it didn't seem possible to predict whether the album or the single gets picked up first (sometimes it's former and sometimes it's the latter).

I have reviewed this and found I am telling lies here: the import sequence is predictable - the importer picks up whichever is first in the alphabet, in this case album.

I have now made this a list.

File: test/plugins/test_hook.py
Comment (snejus) [reply test/plugins/test_hook.py:61]: I think this kind of improvement may be out of scope of the structural change I did here

File: test/plugins/test_hook.py
Comment (snejus) [reply test/plugins/test_hook.py:110]: Nope, Python won't accept two arguments that have the same name: 
image

In this case we want to scrap all arguments, so I think `*_ / *args` may be the most common way to do so: https://www.perplexity.ai/search/python-ignore-positional-argum-f9pNbJEiR76a6eVlJKrgPA

File: test/plugins/test_hook.py
Comment (snejus) [reply test/plugins/test_hook.py:77]: Added a docstring with some details

**Comment by snejus** [comment]
I'm back! @bal-e addressed your comments!

File: test/plugins/test_hook.py
Comment (bal-e) [reply test/plugins/test_hook.py:61]: Alright, could you leave a TODO comment so that somebody can pick it up in the future?

File: test/plugins/test_hook.py
Comment (bal-e) [reply test/plugins/test_hook.py:122]: Alright, cool.  Leave it as it is.

File: test/plugins/test_hook.py
Comment (snejus) [reply test/plugins/test_hook.py]: Had a quick look into this and got it done!

**Review by bal-e — APPROVED** [review]
Amazing work @snejus!

**Comment by Serene-Arc** [comment]
I have no objections! Merge whenever you'd like @snejus

**Pull request merged**
